### ISSUE TITLE
Link-local IPv6Address with scope_id cannot be exploded

### ISSUE THREAD

**Issue by Jonny007-MKD** [pydantic/pydantic#8290]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

When trying to explode a link-local address having a scope identifier (e.g. interface name), I get the following error:

    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/usr/lib/python3.11/ipaddress.py", line 392, in exploded
        return self._explode_shorthand_ip_string()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/ipaddress.py", line 1838, in _explode_shorthand_ip_string
       ip_int = self._ip_int_from_string(ip_str)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/ipaddress.py", line 1719, in _ip_int_from_string
        raise AddressValueError("%s in %r" % (exc, ip_str)) from None
    ipaddress.AddressValueError: Only hex digits permitted in 'ef11%wlo1' in 'fe80::1234:5678:abcd:ef11%wlo1'

Exploding IPv6Addresses should be possible even with scope ids, and the scope id should be appended to the exploded address.

### Example Code

```Python
from pydantic.networks import IPv6Address
a = IPv6Address("fe80::1234:5678:abcd:ef11%wlo1")
print(a.scope_id)
print(a.exploded)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.5.2
        pydantic-core version: 2.14.5
          pydantic-core build: profile=release pgo=true
                 install path: /tmp/bla/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
                     platform: Linux-6.1.0-13-amd64-x86_64-with-glibc2.36
             related packages: typing_extensions-4.8.0
```

**Comment by Jonny007-MKD** [comment]
This is the same with v1.10.4. Shall I open a separate issue for that?

**Comment by sydney-runkle** [comment]
@Jonny007-MKD,

This is due to the fact that `pydantic` uses `IPv6Address` from the stdlib `ipaddress` module, which does not support exploding for addresses with scope IDs. Best case would be to submit a PR and get this fixed in the `ipaddress` module, but a shorter term solution would be to subclass the `IPv6Address` within `pydantic` and add support.

PRs welcome with a fix!

**Comment by NeevCohen** [comment]
Hi @sydney-runkle, I'm new around here and I really love the library and wanted to take a jab at this issue.

Looking in the `ipaddress` module from stdlib, it seems that the issue is in the [`_explode_shorthand_ip_string `](https://github.com/python/cpython/blob/cde141717578f22947553db776980aa3e8801353/Lib/ipaddress.py#L1831) function in the `_BaseV6` class, which doesn't seem to check for a scope id in the `ip_str`.

I followed your advice and tried to subclass the `IPv6Address` class, and check if the `ip_str` contains a scope id using the [`_split_scope_id`](https://github.com/python/cpython/blob/cde141717578f22947553db776980aa3e8801353/Lib/ipaddress.py#L1862) method and call `self._ip_int_from_string(ip_str)` only with the actual IP address, without the scope id, and appending the scope id to the return string.

Something along the lines of

```py
class IPv6AddressFixed(IPv6Address):
    def _explode_shorthand_ip_string(self):
        """Expand a shortened IPv6 address.

        Returns:
            A string, the expanded IPv6 address.

        """
        if isinstance(self, IPv6Network):
            ip_str = str(self.network_address)
        elif isinstance(self, IPv6Interface):
            ip_str, scope_id = self._split_scope_id(str(self.ip))
        else:
            ip_str, scope_id = self._split_scope_id(str(self))

        ip_int = self._ip_int_from_string(self._split_scope_id(ip_str)[0])
        hex_str = '%032x' % ip_int
        parts = [hex_str[x:x+4] for x in range(0, 32, 4)]
        if isinstance(self, (_BaseNetwork, IPv6Interface)):
            return '%s/%d' % (':'.join(parts), self._prefixlen)
        return '%s%%%s' % (':'.join(parts), scope_id)
```

Running the following

```py
print(IPv6AddressFixed('fe80::1%eth0').exploded)
```

outputs 

```
fe80:0000:0000:0000:0000:0000:0000:0001%eth0
```

Does this seem like a good fix? There are references to some private methods and classes like `_BaseNetwork` which I'm not sure are a good idea, but that is the logic in the original method.

**Comment by sydney-runkle** [comment]
@NeevCohen,

So glad you're helping with this 😄.

Seems like a great start. Could you open a PR with these changes? I'd be happy to review and provide some more in-depth feedback.

**Comment by NeevCohen** [comment]
@sydney-runkle So I came up with this solution 

```py
class IPv6Address(IPv6Address):
    @property
    def exploded(self):
        ip_str, scope_id = self._split_scope_id(str(self))
        if not scope_id:
            return super().exploded
        return f'{IPv6Address(ip_str).exploded}%{scope_id}'

    @staticmethod
    def _split_scope_id(ip_str):
        """Helper function to parse IPv6 string address with scope id.

        See RFC 4007 for details.

        Args:
            ip_str: A string, the IPv6 address.

        Returns:
            (addr, scope_id) tuple.

        """
        addr, sep, scope_id = ip_str.partition('%')
        if not sep:
            scope_id = None
        elif not scope_id or '%' in scope_id:
            raise AddressValueError(f'Invalid IPv6 address: "{ip_str}"')
        return addr, scope_id
```

My problem is when running the type checker with `make`, I get the following error 

```
/Users/neevcohen/work/pydantic/pydantic/networks.py
  /Users/neevcohen/work/pydantic/pydantic/networks.py:12:5 - error: Expression of type "type[ipaddress.IPv6Address]" cannot be assigned to declared type "type[pydantic.networks.IPv6Address]"
    "ipaddress.IPv6Address" is incompatible with "pydantic.networks.IPv6Address"
    Type "type[ipaddress.IPv6Address]" cannot be assigned to type "type[pydantic.networks.IPv6Address]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations

make: *** [typecheck] Error 1
```

**Comment by Jonny007-MKD** [comment]
No, probably you should not use a different name for your class, but import `ipaddress.IPv6Address` under a different name. This way the code of pydantic users remains compatible, i.e. they can continue to use `pydantic.networks.IPv6Address`.

I don't know whether pydantic has some principals for doing this, but I would change line 7 like this:

    from ipaddress import IPv4Address, IPv4Interface, IPv4Network, IPv6Address as _IPv6Address, IPv6Interface, IPv6Network

and than use that as base class for the new class. 

    class IPv6Address(_IPv6Address):
        ...

Maybe you also want to add a test in https://github.com/pydantic/pydantic/blob/main/tests/test_networks_ipaddress.py to check whether the new explode is doing what it's supposed to do? I don't know pytest, but probably it could look like this:

    @pytest.mark.parametrize(
        'input','exploded',
        [
            "ff01::2", "ff01:0000:0000:0000:0000:0000:0000:0002",
            "ff01::1%scope", "ff01:0000:0000:0000:0000:0000:0000:0001%scope",
            "1111:2222:3333:4444:5555:6666:7777:8888%test", "1111:2222:3333:4444:5555:6666:7777:8888%test",
        ],
    )
    def test_ipv6address_explode(input, exploded):
        assert IPv6Address(input).exploded == exploded

**Comment by NeevCohen** [comment]
@Jonny007-MKD yeah that seems like a great solution. I'll change it and submit a PR. Thanks!

**Comment by NeevCohen** [comment]
@sydney-runkle I opened a PR 😄

**Comment by sydney-runkle** [comment]
Closing this as not planned, as explained in https://github.com/pydantic/pydantic/pull/8366#issuecomment-1855936151.

TLDR - the onus of fixing this lies on `ipaddress`. However, we're excited that @NeevCohen and @Jonny007-MKD have proposed a workable solution that can cleanly reside in user code 🚀.

**Comment by Jonny007-MKD** [comment]
An issue already exists at Python since 20 months:
https://bugs.python.org/issue44012
https://github.com/python/cpython/issues/88178
